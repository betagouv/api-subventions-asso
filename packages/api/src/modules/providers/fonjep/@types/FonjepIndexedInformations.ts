export default interface FonjepIndexedInformations {
    updated_at: Date;
    unique_id: string;
    code_poste: string;
    montant_paye: number;
    status: string;
    raison: string;
    plein_temps: string;
    service_instructeur: string;
    annee_demande: number;
    date_fin_triennale: Date;
    type_post: string;
    ville: string;
    code_postal?: string;
    contact: string;
    dispositif: string;
    joinKey: string;
}
