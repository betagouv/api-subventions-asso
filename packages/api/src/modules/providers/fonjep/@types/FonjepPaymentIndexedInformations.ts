export default interface FonjepPaymentIndexedInformations {
    unique_id: string;
    updated_at: Date;
    code_poste: string;
    periode_debut: Date;
    periode_fin: Date;
    date_versement: Date;
    montant_a_payer: number;
    montant_paye: number;
    bop: number;
    joinKey: string;
}
