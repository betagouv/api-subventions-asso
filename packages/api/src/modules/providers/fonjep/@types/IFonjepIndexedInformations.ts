export default interface IFonjepIndexedInformations {
    updated_at: Date,
    unique_id: string,
    montant_paye: number,
    status: string,
    service_instructeur: string,
    annee_demande: number,
    date_fin_triennale: Date,
    type_post: string,
    ville: string,
    code_postal?: string,
    contact: string,
    co_financeur?: string,
    co_financeur_contact?: string,
    co_financeur_siret?: string,
    co_financeur_montant?: number,
}
