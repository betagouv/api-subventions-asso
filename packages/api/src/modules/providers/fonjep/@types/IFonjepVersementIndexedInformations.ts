export default interface IFonjepVersementIndexedInformations {
    unique_id: string;
    updated_at: Date;
    code_poste: string;
    periode_debut: Date;
    periode_fin: Date;
    date_versement: Date;
    montant_a_payer: number;
    montant_paye: number;
    bop: string | number;
}
