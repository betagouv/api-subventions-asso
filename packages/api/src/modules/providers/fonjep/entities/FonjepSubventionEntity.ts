import * as ParseHelper from "../../../../shared/helpers/ParserHelper";
import { Siret } from "@api-subventions-asso/dto";
import { ParserInfo, ParserPath, DefaultObject } from "../../../../@types";
import IFonjepIndexedInformations from "../@types/IFonjepIndexedInformations";
import { formatCP } from "../../../../shared/helpers/DataFormatHelper";

export default class FonjepSubventionEntity {
    public static indexedLegalInformationsPath: DefaultObject<ParserPath | ParserInfo> = {
        siret: {
            path: ["Association", "SiretOuRidet"],
            adapter: value => {
                if (!value) return value;

                return value.replace(/ /g, "");
            }
        },
        name: ["Association", "RaisonSociale"]
    };

    public static indexedProviderInformationsPath: DefaultObject<ParserPath | ParserInfo> = {
        code_poste: ["Code"],
        dispositif: ["Dispositif", "Libelle"],
        montant_paye: {
            path: ["MontantSubvention"],
            adapter: value => {
                if (!value) return 0;

                return !value.length ? parseFloat(value) : 0;
            }
        },
        status: ["PstStatutPosteLibelle"],
        raison: ["PstRaisonStatutLibelle"],
        service_instructeur: ["Financeur", "RaisonSociale"],
        annee_demande: ["Annee"],
        date_fin_triennale: {
            path: ["DateFinTriennalite"],
            adapter: value => {
                if (!value) return value;
                return ParseHelper.ExcelDateToJSDate(Number(value));
            }
        },
        updated_at: ["updated_at"],
        unique_id: ["id"],
        type_post: ["TypePoste", "Libelle"],
        ville: ["Association", "Ville"],
        code_postal: {
            path: ["Association", "CodePostal"],
            adapter: formatCP
        },
        contact: ["Association", "ContactEmail"],
        plein_temps: ["PleinTemps"]
    };

    constructor(
        public legalInformations: {
            siret: Siret;
            name: string;
        },
        public indexedInformations: IFonjepIndexedInformations,
        public data: unknown
    ) {}
}
