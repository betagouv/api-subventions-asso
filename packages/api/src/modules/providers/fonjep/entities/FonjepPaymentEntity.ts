import { Siret } from "dto";
import { DefaultObject, ParserInfo, ParserPath } from "../../../../@types";
import IFonjepVersementIndexedInformations from "../@types/IFonjepPaymentIndexedInformations";
import * as ParseHelper from "../../../../shared/helpers/ParserHelper";

export default class FonjepPaymentEntity {
    public static indexedLegalInformationsPath: DefaultObject<ParserPath | ParserInfo> = {
        siret: {
            path: ["siret"],
            adapter: value => {
                if (!value) return value;
                return value.replace(/ /g, "");
            },
        },
    };

    public static indexedProviderInformationsPath: DefaultObject<ParserPath | ParserInfo> = {
        // TODO <string|number>
        unique_id: ["id"],
        updated_at: ["updated_at"],
        code_poste: ["PosteCode"],
        bop: ["bop"],
        periode_debut: {
            path: ["PeriodeDebut"],
            adapter: value => {
                if (!value) return value;
                return ParseHelper.ExcelDateToJSDate(Number(value));
            },
        },
        periode_fin: {
            path: ["PeriodeFin"],
            adapter: value => {
                if (!value) return value;
                return ParseHelper.ExcelDateToJSDate(Number(value));
            },
        },
        date_versement: {
            path: ["DateVersement"],
            adapter: value => {
                if (!value) return value;
                return ParseHelper.ExcelDateToJSDate(Number(value));
            },
        },
        montant_a_payer: ["MontantAPayer"],
        montant_paye: ["MontantPaye"],
    };

    constructor(
        public legalInformations: {
            siret: Siret;
        },
        public indexedInformations: IFonjepVersementIndexedInformations,
        public data: unknown,
    ) {}
}
