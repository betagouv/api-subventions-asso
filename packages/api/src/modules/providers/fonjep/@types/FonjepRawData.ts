import FonjepSubventionEntity from "../entities/FonjepSubventionEntity";
import FonjepVersementEntity from "../entities/FonjepVersementEntity";

export interface FonjepRowData {
    subvention: FonjepSubventionEntity;
    versements: FonjepVersementEntity[];
}
