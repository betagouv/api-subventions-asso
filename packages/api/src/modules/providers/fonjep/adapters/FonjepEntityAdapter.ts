import FonjepTiersEntity from "../entities/FonjepTiersEntity";
import FonjepPosteEntity from "../entities/FonjepPosteEntity";
import FonjepVersementEntity, { PayedFonjepVersementEntity } from "../entities/FonjepVersementEntity";
import FonjepTypePosteEntity from "../entities/FonjepTypePosteEntity";
import FonjepDispositifEntity from "../entities/FonjepDispositifEntity";
import FonjepDispositifDto from "../../../../dataProviders/db/providers/fonjep/dto/fonjepDispositifDto";
import { FonjepPosteDtoWithJSDate } from "../../../../dataProviders/db/providers/fonjep/dto/fonjepPosteDto";
import FonjepTypePosteDto from "../../../../dataProviders/db/providers/fonjep/dto/fonjepTypePosteDto";
import { FonjepVersementDto } from "../../../../dataProviders/db/providers/fonjep/dto/fonjepVersementDto";
import FonjepTiersDto from "../../../../dataProviders/db/providers/fonjep/dto/fonjepTiersDto";
import Siret from "../../../../identifierObjects/Siret";
import Ridet from "../../../../identifierObjects/Ridet";
import Rid from "../../../../identifierObjects/Rid";
import Siren from "../../../../identifierObjects/Siren";
import { GenericAdapter } from "../../../../shared/GenericAdapter";
import { FonjepApplicationFlatEntity, FonjepPaymentFlatEntity } from "../entities/FonjepFlatEntity";
import { DataBretagneRecords } from "../../dataBretagne/@types/DataBretagne";
import dataBretagneService from "../../dataBretagne/dataBretagne.service";
import { getShortISODate, modifyDateYear } from "../../../../shared/helpers/DateHelper";
import { removeWhitespace } from "../../../../shared/helpers/StringHelper";
import EstablishmentIdentifier from "../../../../identifierObjects/EstablishmentIdentifier";
import { ApplicationNature, ApplicationStatus, PaymentCondition } from "dto";

/**
 * Some of the nullIfEmpty calls have not been verified and were added base on every FonjepEntity type
 * Properties without nullIfEmpty are supposed to be mandatory / always present (checked in the 2024-12-31 extract file)
 */
export default class FonjepEntityAdapter {
    // duplicate with fonjepService.provider.id
    // see https://github.com/betagouv/api-subventions-asso/issues/3256
    static PROVIDER_NAME = "Fonjep";

    /**
     * This was defined a long time ago and must be updated and validated
     */
    private static FOUNDER_CODE_TO_BOP_MAPPER = {
        "10004": 163,
        "10005": 163,
        "10008": 147,
        "10009": 163,
        "10010": 209,
        "10012": 361,
        "10016": 163,
        "10017": 163,
    };

    /**
     *
     * @param code Financer code
     * It can be found in :
     * financeurCofinanceurCode (fonjepVersement)
     * financeurCode (fonjepDispositif)
     * financeurPrincipalCode (fonjepPoste)
     * code (fonjepTiers)
     *
     * @returns The state financial program number
     */
    static getBopFromFounderCode(code: number): number {
        return this.FOUNDER_CODE_TO_BOP_MAPPER[code];
    }

    static toFonjepTierEntity(tier: FonjepTiersDto, updateDate: Date): FonjepTiersEntity {
        return {
            code: tier["Code"],
            raisonSociale: tier["RaisonSociale"],
            estAssociation: tier["EstAssociation"],
            estCoFinanceurPostes: tier["EstCoFinanceurPostes"],
            estFinanceurPostes: tier["EstFinanceurPostes"],
            siretOuRidet: tier["SiretOuRidet"] ? removeWhitespace(tier["SiretOuRidet"]) : null,
            codePostal: tier["CodePostal"],
            ville: tier["Ville"],
            contactEmail: tier["ContactEmail"],
            updateDate,
        };
    }

    static toFonjepPosteEntity(poste: FonjepPosteDtoWithJSDate, updateDate: Date): FonjepPosteEntity {
        return {
            code: poste["Code"],
            annee: poste["Annee"],
            associationBeneficiaireCode: poste["AssociationBeneficiaireCode"],
            financeurPrincipalCode: poste["FinanceurPrincipalCode"],
            financeurAttributeurCode: poste["FinanceurAttributeurCode"],
            dateFinTriennalite: poste["DateFinTriennalite"],
            dispositifId: poste["DispositifId"],
            pstStatutPosteLibelle: poste["PstStatutPosteLibelle"],
            pstRaisonStatutLibelle: poste["PstRaisonStatutLibelle"],
            associationImplantationCode: poste["AssociationImplantationCode"],
            montantSubvention: poste["MontantSubvention"],
            pstTypePosteCode: poste["PstTypePosteCode"],
            pleinTemps: poste["PleinTemps"],
            doublementUniteCompte: poste["DoublementUniteCompte"],
            updateDate,
        };
    }

    static toFonjepVersementEntity(versement: FonjepVersementDto, updateDate: Date): FonjepVersementEntity {
        return {
            posteCode: versement["PosteCode"],
            periodeDebut: versement["PeriodeDebut"],
            periodeFin: versement["PeriodeFin"],
            dateVersement: versement["DateVersement"],
            montantAPayer: versement["MontantAPayer"],
            montantPaye: versement["MontantPaye"],
            updateDate,
        };
    }

    static toFonjepTypePosteEntity(typePoste: FonjepTypePosteDto, updateDate: Date): FonjepTypePosteEntity {
        return {
            code: typePoste["Code"],
            libelle: typePoste["Libelle"],
            updateDate,
        };
    }

    static toFonjepDispositifEntity(dispositif: FonjepDispositifDto, updateDate: Date): FonjepDispositifEntity {
        return {
            id: dispositif["ID"],
            libelle: dispositif["Libelle"],
            financeurCode: dispositif["FinanceurCode"],
            updateDate,
        };
    }

    public static extractPositionCode(entity: FonjepPaymentFlatEntity) {
        if (entity.ej !== GenericAdapter.NOT_APPLICABLE_VALUE) {
            throw new Error("You must extract a position code from a FonjepPaymentFlat entity");
        } else {
            // cf buildFlatPaymentId
            return entity.idVersement.split("-")[0];
        }
    }

    private static getConventionDate(position: FonjepPosteEntity): Date {
        if (position.dateFinTriennalite) return modifyDateYear(position.dateFinTriennalite, -3);
        throw new Error("We can't create FONJEP PaymentFlat without dateFinTriennalite");
    }

    private static buildFlatPaymentId(data: { thirdParty: FonjepTiersEntity; position: FonjepPosteEntity }) {
        const { thirdParty, position } = data;
        return `${position.code}-${getShortISODate(FonjepEntityAdapter.getConventionDate(position))}-${position.annee}-${
            thirdParty.siretOuRidet
        }`;
    }

    // this keeps the same structure as other providers payment flat uniqueId and add N/A for the missing fields
    private static buildPaymentFlatUniqueId(
        partialPaymentFlat: Omit<FonjepPaymentFlatEntity, "uniqueId">,
        beginningDate: Date,
    ) {
        const keys = [
            "fonjep",
            partialPaymentFlat.idVersement,
            partialPaymentFlat.programNumber,
            GenericAdapter.NOT_APPLICABLE_VALUE, // action code
            GenericAdapter.NOT_APPLICABLE_VALUE, // activity code
            getShortISODate(beginningDate), // was operationDate to be ISO with CHORUS but it can be identical for every payment of the same year. See positionCode S05543
            GenericAdapter.NOT_APPLICABLE_VALUE, // attachement comptable
            GenericAdapter.NOT_APPLICABLE_VALUE, // centre financier code
        ];
        return keys.join("-");
    }

    static toFonjepPaymentFlat(
        fonjepData: { payment: PayedFonjepVersementEntity; position: FonjepPosteEntity; thirdParty: FonjepTiersEntity },
        dataBretagneData: DataBretagneRecords,
    ): FonjepPaymentFlatEntity {
        const { payment, thirdParty, position } = fonjepData;
        if (!thirdParty.siretOuRidet) {
            throw new Error("Trying to create a FONJEP PaymentFlat without siret or ridet information");
        } else {
            const estabId = thirdParty.siretOuRidet;
            let estabIdType: "siret" | "ridet";
            let assoIdType: "siren" | "rid";
            let estabValueObjectId: Siret | Ridet;
            let assoValueObjectId: Siren | Rid;

            if (Siret.isSiret(estabId)) {
                estabIdType = "siret";
                assoIdType = "siren";
                estabValueObjectId = new Siret(estabId);
                assoValueObjectId = estabValueObjectId.toSiren();
            } else if (Ridet.isRidet(estabId)) {
                estabIdType = "ridet";
                assoIdType = "rid";
                estabValueObjectId = new Ridet(estabId);
                assoValueObjectId = estabValueObjectId.toRid();
            } else {
                throw new Error(
                    `Fonjep Tier with code ${position.code} has no siret or ridet and should not have been parsed`,
                );
            }

            const program =
                dataBretagneData.programs[
                    this.getBopFromFounderCode(Number(fonjepData.position.financeurPrincipalCode))
                ];

            const ministry = dataBretagneService.getMinistryEntity(program, dataBretagneData.ministries);

            const paymentId = this.buildFlatPaymentId({
                thirdParty,
                position,
            });

            const partialPaymentFlat = {
                idVersement: paymentId,
                exerciceBudgetaire: position.annee as number,
                typeIdEtablissementBeneficiaire: estabIdType,
                idEtablissementBeneficiaire: estabValueObjectId,
                typeIdEntrepriseBeneficiaire: assoIdType,
                idEntrepriseBeneficiaire: assoValueObjectId,
                amount: payment.montantPaye,
                operationDate: payment.dateVersement,
                centreFinancierCode: GenericAdapter.NOT_APPLICABLE_VALUE,
                centreFinancierLibelle: GenericAdapter.NOT_APPLICABLE_VALUE,
                attachementComptable: GenericAdapter.NOT_APPLICABLE_VALUE,
                regionAttachementComptable: GenericAdapter.NOT_APPLICABLE_VALUE,
                ej: GenericAdapter.NOT_APPLICABLE_VALUE,
                actionCode: GenericAdapter.NOT_APPLICABLE_VALUE,
                actionLabel: GenericAdapter.NOT_APPLICABLE_VALUE,
                activityCode: GenericAdapter.NOT_APPLICABLE_VALUE,
                activityLabel: GenericAdapter.NOT_APPLICABLE_VALUE,
                provider: this.PROVIDER_NAME.toLowerCase(),
                programName: program.label_programme,
                programNumber: program.code_programme,
                mission: program.mission,
                ministry: ministry?.nom_ministere || null,
                ministryAcronym: ministry?.sigle_ministere || null,
                updateDate: payment.updateDate,
            };

            // TODO: Another example where something nullable (Tier.FinanceurPrincipalCode) is required to build a unique ID
            // TODO: make FonjepTierEntity.financeurPrincipalCode mandatory and make it a number
            const uniqueId = this.buildPaymentFlatUniqueId(partialPaymentFlat, payment.periodeDebut);

            return { uniqueId, ...partialPaymentFlat };
        }
    }

    static toFonjepApplicationFlat(entities: {
        position: FonjepPosteEntity;
        beneficiary: FonjepTiersEntity; // association who made the application
        allocator: FonjepTiersEntity | undefined; // structure who handle the application payment
        instructor: FonjepTiersEntity | undefined; // structure who validates / instructs the application
        scheme: FonjepDispositifEntity | undefined;
    }): Omit<FonjepApplicationFlatEntity, "updateDate"> | null {
        const { position, beneficiary, allocator, instructor, scheme } = entities;
        if (!position.annee) throw new Error("FONJEP ApplicationFlat must have a budgetary year");
        if (!beneficiary.siretOuRidet) throw new Error("FONJEP ApplicationFlat must have a beneficiary siret or ridet");
        const beneficiaryEstablishmentIdType =
            EstablishmentIdentifier.getIdentifierType(beneficiary.siretOuRidet) || null;
        if (!beneficiaryEstablishmentIdType) {
            // TODO: this should not happen but has been encountered where rid was present in instead of ridet. See #3586
            console.log(
                `FONJEP ApplicationFlat must have a valid beneficiary siret or ridet, given ${beneficiary.siretOuRidet}.`,
            );
            return null;
        }

        const provider = this.PROVIDER_NAME.toLowerCase(); // replace this with #3338
        const applicationProviderId = `${position.code}-${getShortISODate(this.getConventionDate(position))}`;
        const applicationId = `${provider}-${applicationProviderId}`;
        const uniqueId = `${applicationId}-${position.annee}`;
        const paymentId = this.buildFlatPaymentId({ thirdParty: beneficiary, position });

        const allocatorIdType = allocator?.siretOuRidet
            ? EstablishmentIdentifier.getIdentifierType(allocator.siretOuRidet)
            : null;
        const instructiveDepartmentIdType = instructor?.siretOuRidet
            ? EstablishmentIdentifier.getIdentifierType(instructor.siretOuRidet)
            : null;

        return {
            uniqueId,
            applicationId,
            applicationProviderId,
            provider,
            joinKeyId: GenericAdapter.NOT_APPLICABLE_VALUE,
            joinKeyDesc: GenericAdapter.NOT_APPLICABLE_VALUE,
            allocatorName: allocator?.raisonSociale ?? null,
            allocatorId: allocator?.siretOuRidet ?? null,
            allocatorIdType,
            managingAuthorityName: "FONJEP",
            managingAuthorityId: null,
            managingAuthorityIdType: null,
            instructiveDepartmentName: instructor?.raisonSociale ?? null,
            instructiveDepartmentIdType,
            instructiveDepartementId: instructor?.siretOuRidet ?? null,
            beneficiaryEstablishmentId: beneficiary.siretOuRidet,
            beneficiaryEstablishmentIdType,
            budgetaryYear: position.annee,
            pluriannual: true,
            pluriannualYears: null, // null for now, see #3575 for updates
            decisionDate: null,
            conventionDate: this.getConventionDate(position),
            decisionReference: null,
            depositDate: null,
            requestYear: null,
            scheme: scheme?.libelle ?? null,
            subScheme: GenericAdapter.NOT_APPLICABLE_VALUE,
            statusLabel: ApplicationStatus.GRANTED, // always GRANTED because FONJEP only concerns payments (means that application has been granted)
            object: null,
            nature: ApplicationNature.MONEY,
            requestedAmount: position.montantSubvention,
            grantedAmount: position.montantSubvention,
            totalAmount: position.montantSubvention,
            ej: GenericAdapter.NOT_APPLICABLE_VALUE,
            paymentId,
            paymentCondition: PaymentCondition.PHASED,
            paymentConditionDesc: null,
            paymentPeriodDates: null,
            cofinancingRequested: GenericAdapter.NOT_APPLICABLE_VALUE,
            cofinancersNames: GenericAdapter.NOT_APPLICABLE_VALUE,
            cofinancersIdType: GenericAdapter.NOT_APPLICABLE_VALUE,
            confinancersId: GenericAdapter.NOT_APPLICABLE_VALUE,
            idRAE: null,
            ueNotification: null,
            subventionPercentage: null,
        };
    }
}
