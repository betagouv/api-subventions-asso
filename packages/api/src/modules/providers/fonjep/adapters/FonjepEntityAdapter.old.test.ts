import FonjepEntityAdapter from "./FonjepEntityAdapter.old";
import { SubventionEntity, PaymentEntity } from "../../../../../tests/modules/providers/fonjep/__fixtures__/entity";
import ProviderValueFactory from "../../../../shared/ProviderValueFactory";
import { FONJEP_PAYMENTS, FONJEP_PAYMENT_ENTITIES } from "../__fixtures__old/FonjepEntities";
import { RawApplication, RawFullGrant, RawPayment } from "../../../grant/@types/rawGrant";
import { DemandeSubvention, Grant, Payment } from "dto";
import FonjepSubventionEntity from "../entities/FonjepSubventionEntity.old";
import FonjepPaymentEntity from "../entities/FonjepPaymentEntity.old";
import dataBretagneService from "../../dataBretagne/dataBretagne.service";
import PROGRAMS from "../../../../../tests/dataProviders/db/__fixtures__/stateBudgetProgram";

describe("FonjepEntityAdapter", () => {
    beforeAll(() => {
        dataBretagneService.programsByCode = {
            // FONJEP PROGRAM 163
            [PROGRAMS[2].code_programme]: PROGRAMS[2],
        };
        jest.useFakeTimers().setSystemTime(new Date("2022-01-01"));
    });

    describe("toDemandeSubvention()", () => {
        const buildProviderValueAdapterMock = jest.spyOn(ProviderValueFactory, "buildProviderValueAdapter");
        it("should return a DemandeSubvention", () => {
            // @ts-expect-error: mock
            buildProviderValueAdapterMock.mockImplementationOnce(() => value => value);
            const actual = FonjepEntityAdapter.toDemandeSubvention(SubventionEntity);
            expect(actual).toMatchSnapshot();
        });
    });

    describe("rawToGrant", () => {
        // @ts-expect-error: parameter type
        const FONJEP_APPLICATION: FonjepSubventionEntity = { fonjep: "application" };
        // @ts-expect-error: parameter type
        const FONJEP_PAYMENTS: FonjepPaymentEntity[] = [{ fonjep: "payment_1" }, { fonjep: "payment_2" }];
        // @ts-expect-error: parameter type
        const RAW_FULLGRANT: RawFullGrant<{ application: FonjepSubventionEntity; payments: FonjepPaymentEntity[] }> = {
            data: {
                application: FONJEP_APPLICATION,
                payments: FONJEP_PAYMENTS,
            },
        };
        // @ts-expect-error: parameter type
        const APPLICATION: DemandeSubvention = { foo: "bar" };
        // @ts-expect-error: parameter type
        const PAYMENT_1: Payment = { poo: "par" };
        // @ts-expect-error: parameter type
        const PAYMENT_2: Payment = { pee: "pez" };
        const GRANT: Grant = { application: APPLICATION, payments: [PAYMENT_1, PAYMENT_2] };

        let mockToDemandeSubvention: jest.SpyInstance;
        let mockToPayment: jest.SpyInstance;

        beforeAll(() => {
            mockToDemandeSubvention = jest.spyOn(FonjepEntityAdapter, "toDemandeSubvention");
            mockToDemandeSubvention.mockReturnValue(APPLICATION);
            mockToPayment = jest.spyOn(FonjepEntityAdapter, "toPayment");
        });

        beforeEach(() => {
            mockToPayment.mockReturnValueOnce(PAYMENT_1);
            mockToPayment.mockReturnValueOnce(PAYMENT_2);
        });

        afterAll(() => {
            mockToDemandeSubvention.mockRestore();
            mockToPayment.mockRestore();
        });

        it("should call toDemandeSubvention", () => {
            FonjepEntityAdapter.rawToGrant(
                RAW_FULLGRANT,
                FONJEP_PAYMENTS.map(() => PROGRAMS[1]),
            );
            expect(mockToDemandeSubvention).toHaveBeenCalledWith(FONJEP_APPLICATION);
        });

        it("should call toPayment", () => {
            FonjepEntityAdapter.rawToGrant(
                RAW_FULLGRANT,
                FONJEP_PAYMENTS.map(() => PROGRAMS[1]),
            );
            expect(mockToPayment).toHaveBeenCalledTimes(FONJEP_PAYMENTS.length);
            FONJEP_PAYMENTS.forEach((payment, index) => {
                expect(mockToPayment).toHaveBeenNthCalledWith(index + 1, payment, PROGRAMS[1]);
            });
        });

        it("should return Grant", () => {
            const expected = GRANT;
            const actual = FonjepEntityAdapter.rawToGrant(
                RAW_FULLGRANT,
                FONJEP_PAYMENTS.map(() => PROGRAMS[1]),
            );
            expect(actual).toEqual(expected);
        });
    });

    describe("rawToApplication", () => {
        // @ts-expect-error: parameter type
        const RAW_APPLICATION: RawApplication<FonjepSubventionEntity> = { data: { foo: "bar" } };
        // @ts-expect-error: parameter type
        const APPLICATION: DemandeSubvention = { foo: "bar" };
        let mockToDemandeSubvention: jest.SpyInstance;

        beforeAll(() => {
            mockToDemandeSubvention = jest.spyOn(FonjepEntityAdapter, "toDemandeSubvention");
            mockToDemandeSubvention.mockReturnValue(APPLICATION);
        });

        afterAll(() => {
            mockToDemandeSubvention.mockRestore();
        });

        it("should call toDemandeSubvention", () => {
            FonjepEntityAdapter.rawToApplication(RAW_APPLICATION);
            expect(mockToDemandeSubvention).toHaveBeenCalledWith(APPLICATION);
        });

        it("should return DemandeSubvention", () => {
            const expected = APPLICATION;
            const actual = FonjepEntityAdapter.rawToApplication(RAW_APPLICATION);
            expect(actual).toEqual(expected);
        });
    });

    describe("rawToPayment", () => {
        // @ts-expect-error: parameter type
        const RAW_PAYMENT: RawPayment<FonjepPaymentEntity> = { data: FONJEP_PAYMENT_ENTITIES[0] };
        let mockToPayment: jest.SpyInstance;

        beforeAll(() => {
            mockToPayment = jest.spyOn(FonjepEntityAdapter, "toPayment");
            mockToPayment.mockReturnValue(FONJEP_PAYMENTS[0]);
        });

        afterAll(() => {
            mockToPayment.mockRestore();
        });

        it("should call toPayment", () => {
            FonjepEntityAdapter.rawToPayment(RAW_PAYMENT, PROGRAMS[1]);
            expect(FonjepEntityAdapter.toPayment).toHaveBeenCalledWith(RAW_PAYMENT.data, PROGRAMS[1]);
        });

        it("should return Payment", () => {
            const expected = FONJEP_PAYMENTS[0];
            const actual = FonjepEntityAdapter.rawToPayment(RAW_PAYMENT, PROGRAMS[1]);
            expect(actual).toEqual(expected);
        });
    });

    describe("toPayment", () => {
        const buildProviderValueAdapterMock = jest.spyOn(ProviderValueFactory, "buildProviderValueAdapter");

        it("should return a Payment", () => {
            // @ts-expect-error: mock
            buildProviderValueAdapterMock.mockImplementationOnce(() => value => value);
            const actual = FonjepEntityAdapter.toPayment(PaymentEntity, PROGRAMS[1]);
            expect(actual).toMatchSnapshot();
        });
    });

    describe("toEtablissement()", () => {
        const buildProviderValuesAdapterMock = jest.spyOn(ProviderValueFactory, "buildProviderValuesAdapter");
        it("should return an Etablissement", () => {
            // @ts-expect-error: mock
            buildProviderValuesAdapterMock.mockImplementationOnce(() => value => [value]);
            const actual = FonjepEntityAdapter.toEtablissement(SubventionEntity);
            expect(actual).toMatchSnapshot();
        });
    });

    describe("toCommon", () => {
        it("returns proper result", () => {
            const INPUT = {
                indexedInformations: {
                    date_versement: new Date("2022-02-02"),
                    dispositif: "DISPOSITIF",
                    annee_demande: 2023,
                    montant_paye: 42,
                    service_instructeur: "INSTRUCTEUR",
                },
                legalInformations: { siret: "123456789" },
                data: { MontantSubvention: 43 },
            };
            const actual = FonjepEntityAdapter.toCommon(INPUT);
            expect(actual).toMatchSnapshot();
        });
    });
});
