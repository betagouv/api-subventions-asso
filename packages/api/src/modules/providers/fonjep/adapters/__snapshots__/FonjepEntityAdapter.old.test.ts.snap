// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FonjepEntityAdapter toCommon returns proper result 1`] = `
{
  "bop": "",
  "date_debut": 2022-02-02T00:00:00.000Z,
  "dispositif": "DISPOSITIF",
  "exercice": 2023,
  "montant_accorde": 43,
  "montant_verse": 42,
  "objet": "",
  "service_instructeur": "INSTRUCTEUR",
  "siret": "123456789",
  "statut": "Accordé",
}
`;

exports[`FonjepEntityAdapter toDemandeSubvention() should return a DemandeSubvention 1`] = `
{
  "annee_demande": 2022,
  "date_fin": 2022-01-03T00:00:00.000Z,
  "dispositif": "Dispositif",
  "montants": {
    "accorde": 500,
    "demande": 500,
  },
  "plein_temps": "Oui",
  "pluriannualite": "Oui",
  "service_instructeur": "XXX",
  "siret": "10000000000001",
  "status": "Attribué - Reconduction",
  "statut_label": "Accordé",
  "versementKey": "J00034",
}
`;

exports[`FonjepEntityAdapter toEstablishment() should return an Establishment 1`] = `
{
  "adresse": [
    {
      "code_postal": "75000",
      "commune": "Paris",
    },
  ],
  "contacts": [
    [
      {
        "email": "contact@beta.gouv.fr",
      },
    ],
  ],
  "nic": [
    "00001",
  ],
  "siret": [
    "10000000000001",
  ],
}
`;

exports[`FonjepEntityAdapter toPayment should return a Payment 1`] = `
{
  "amount": 1000,
  "bop": 163,
  "codePoste": "J00034",
  "dateOperation": 2022-01-02T00:00:00.000Z,
  "libelleProgramme": "PROGRAM 163",
  "montantAPayer": 1000,
  "periodeDebut": 2022-01-02T00:00:00.000Z,
  "periodeFin": 2022-01-02T00:00:00.000Z,
  "programme": 163,
  "siret": "10000000000001",
  "versementKey": "J00034",
}
`;
